compiler/util/src/org/jetbrains/kotlin/utils/SmartHashMap.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.utils
+
+const val ARRAY_UNTIL_SIZE = 10
+
+/**
+ * [SmartHashMap] is a Map implementation that uses reference identity for keys.
+ * It uses 2 arrays to store keys & values until the number of entries stored is larger than 10.
+ * At that point it switches to using a HashMap and stays that way until [clear] is called.
+ * It does not auto convert back if the number of entries decreases through [remove].
+ */
+class SmartHashMap<K, V>() : MutableMap<K, V> {
+
+    private var keysArray: MutableList<K>? = arrayListOf()
+    private var valuesArray: MutableList<V>? = arrayListOf()
+    private var largeMap: HashMap<K, V>? = null
+
+    override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
+        get() {
+            return keysArray?.let {
```

Comment:
Look into using `mapTo()` to make this (and `keys`, `values`, and `convertToHashMap()`) more "Kotlin-ish" (or `mapNotNullTo()` if you take my suggestion of using Array. https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/map-not-null-to.html

Reply:
cleaned up a different way